src/LAMMPS.jl

```diff
@@ -177,240 +245,344 @@ function get_natoms(lmp::LMP)
     Int64(API.lammps_get_natoms(lmp))
 end
 
-function dtype2type(dtype::API._LMP_DATATYPE_CONST)
-    if dtype == API.LAMMPS_INT
-        type = Ptr{Int32}
-    elseif dtype == API.LAMMPS_INT_2D
-        type = Ptr{Ptr{Int32}}
-    elseif dtype == API.LAMMPS_INT64
-        type = Ptr{Int64}
-    elseif dtype == API.LAMMPS_INT64_2D
-        type = Ptr{Ptr{Int64}}
-    elseif dtype == API.LAMMPS_DOUBLE
-        type = Ptr{Float64}
-    elseif dtype == API.LAMMPS_DOUBLE_2D
-        type = Ptr{Ptr{Float64}}
-    elseif dtype == API.LAMMPS_STRING
-        type = Ptr{Cchar}
-    else
-        @assert false "Unknown dtype: $dtype"
-    end
-    return type
+function _string(ptr::Ptr)
+    ptr == C_NULL && error("Wrapping NULL-pointer!")
+    return Base.unsafe_string(ptr)
 end
 
-"""
-    extract_global(lmp, name, dtype=nothing)
-"""
-function extract_global(lmp::LMP, name, dtype=nothing)
-    if dtype === nothing
-        dtype = API.lammps_extract_global_datatype(lmp, name)
-    end
-    dtype = API._LMP_DATATYPE_CONST(dtype)
-    type = dtype2type(dtype)
+function _extract(ptr::Ptr{<:Real}, shape::Integer; copy=false, own=false)
```

Comment:
`copy` should imply `own`?  We can almost never use `own=true` since we need to use lammps_free for the deallocation

Reply:
As far as I can tell, we don't need to use `lammps_free` for deallocation. It just calls `free` internally without doing anything else. Thats the same thing that would happen If we take ownership of the memory if I'm not mistaken. I've taken this directly from the lammps source code:
```C++
void lammps_free(void *ptr)
{
  free(ptr);
}
```

The `own` keyword is used for the `extract_variable` method, specifically when it's a Atom style variable:
```julia
  if lmp_variable == VAR_ATOM
      ndata = extract_setting(lmp, "nlocal")
      ptr = _reinterpret(LAMMPS_DOUBLE, void_ptr)
      # lammps expects us to take ownership of the data
      return _extract(ptr, ndata; own=true)
  end
```
because the documentation states:
> For atom-style variables, the return value is a pointer to an allocated block of storage of double of the length
atom->nlocal. Since the data returned are a copy, the location will persist, but its content will not be updated in case the variable is re-evaluated. To avoid a memory leak, this pointer needs to be freed after use in the calling program.

That means that the value is already a copy and copying it again would be superfluous.

I've also (crudely) verified, that this does indeed not cause a memory leak by running the following script:

```julia
lmp = LMP()

command(lmp, """
         atom_modify map yes
         region cell block 0 3 0 3 0 3
         create_box 1 cell
         lattice sc 1
         create_atoms 1 region cell
         mass 1 1
        
         variable var atom id
    """)

for _ in 1:10^7
    extract_variable(lmp, "var", VAR_ATOM)
end
```

If I set `own` to false, my memory usage rises by about 2GB. If I set `own` to true the memory consumption stays stable.
So I'd say its safe to let the GC handle freeing the memory if we're supposed to take ownership.

I have also verified that the data stays allocated even if we close the lammps instance. I've even written this as a test:

```julia
lmp = LMP(["-screen", "none"])
command(lmp, """
    atom_modify map yes
    region cell block 0 3 0 3 0 3
    create_box 1 cell
    lattice sc 1
    create_atoms 1 region cell
    mass 1 1

    variable var atom id
""")

var = extract_variable(lmp, "var", VAR_ATOM)
var_copy = copy(var)
LAMMPS.close!(lmp)

@test var == var_copy
```

In summary, we use `copy=true`, if the ownership stays with LAMMPS but we want a copy and `own=true` if the ownership is handed to us, which is usually because LAMMPS has made a copy for us

Reply:
> t just calls free internally without doing anything else. Thats the same thing that would happen If we take ownership of the memory if I'm not mistaken.

This is sadly not true. In shared libraries the `malloc` and `free` maybe be resolved to different implementations. So you can never rely that the functions are the same in two shared libraries.

> if we're supposed to take ownership.

So that's actually a good question. When are we supposed to take ownership? 

As far as I understood `extract` we should never take ownershi, but if the underlying API is allocating memory for us, then we will need to do a `copy` + `lammps_free`. Which sucks.

Reply:
Okay according to the docs https://docs.lammps.org/Library_objects.html
atom style requires us to call `lammps_free`.

Reply:
I've thought about registering lammps_free as a finalizer on our array. I think that could work?

Reply:
Doing this

```julia
function _extract(ptr::Ptr{<:Real}, shape::Integer; copy=false, own=false)
    ptr == C_NULL && error("Wrapping NULL-pointer!")
    result = Base.unsafe_wrap(Array, ptr, shape; own=false)

    if own
        finalizer(r -> API.lammps_free(ptr), result)
    end

    return copy ? Base.copy(result) : result
end
```
seems to work. It looks like the memory usage flucutaes a bit more this way than simply setting `own=true` for `Base.unsafe_wrap`, but the overall memory usage stays about the same

Reply:
running the test script without free of any sort:

![grafik](https://github.com/cesmix-mit/LAMMPS.jl/assets/32484985/9117e133-7043-4778-bce6-09944bb92177)

running the test script and using `own=true` on `Base.unsafe_wrap`:

![grafik](https://github.com/cesmix-mit/LAMMPS.jl/assets/32484985/c1c0c83d-6223-49f0-ba06-61911922ff0f)

running the test script and registering `lammps_free` as a finalizer:

![grafik](https://github.com/cesmix-mit/LAMMPS.jl/assets/32484985/678446be-ba2a-485e-baa1-ca99d84ebe2f)

Reply:
this is probably better:

```julia
function _extract(ptr::Ptr{<:Real}, shape::Integer; copy=false, own=false)
    ptr == C_NULL && error("Wrapping NULL-pointer!")
    result = Base.unsafe_wrap(Array, ptr, shape; own=false)

    if own
        finalizer(API.lammps_free, result)
    end

    return copy ? Base.copy(result) : result
end
```

I was worried about what happens for resize operations, but julia just doesn't allow them for shared data so that's not a problem

Reply:
That's an interesting idea. I am wondering about `reshape` IIRC there are some operations that return an array again, but yes if the Array is shared, they probably should not return a new Array...

Also 1.11 is introducing `Memory` and I think the finalizer will need to be attached to that.

Reply:
It appears like the finallizer is only run, when the underlying data is no longer acessible. So having a reshaped version of an array available would block the execution of the finalizer´.

Reply:
I've done:
```julia
finalizer(result) do r
    API.lammps_free(result)
    @async println("finializer has run!")
end
```
and then the following test script:

```julia
lmp = LMP()

command(lmp, """
   atom_modify map yes
   region cell block 0 3 0 3 0 3
   create_box 1 cell
   lattice sc 1
   create_atoms 1 region cell
   mass 1 1

   variable var atom id
""")

var1 = extract_variable(lmp, "var", VAR_ATOM)
var2 = reshape(var1, 1, 27)

var1 = nothing
GC.gc() # finallizer is not run

var2 = nothing
GC.gc() # finallizer is run
```

Reply:
I think the finalizer is the best solution. Should we merge this PR?
Then as soon as 1.11 gets released, we should revisit this. It might become necessary to force 1.11 or higher if we need to register the finalizer on the `Memory`?

Reply:
So on 1.11

```
julia> var2 = reshape(var1, 1, 27)
1×27 Matrix{Float64}:
 1.0  2.0  3.0  4.0  5.0  6.0  7.0  8.0  9.0  10.0  11.0  12.0  13.0  14.0  …  17.0  18.0  19.0  20.0  21.0  22.0  23.0  24.0  25.0  26.0  27.0

julia> var1 = nothing

julia> GC.gc() # finallizer is run

julia> var2
1×27 Matrix{Float64}:
 1.0  2.0  3.0  4.0  5.0  6.0  7.0  8.0  9.0  10.0  11.0  12.0  13.0  14.0  …  18.0  19.0  20.0  21.0  22.0  23.0  24.0  25.0  26.0  1.265e-321
```

Note the data flipping.

Reply:
I opened an upstream issue, but I don't expect this behavior to material change before 1.11 gets released

https://github.com/JuliaLang/julia/issues/54985

So if we want to use this finalizer version, we would need to use `@static if VERSION >= v1.11-dev` and attach the finalizer to `result.ref.mem`.

Reply:
this now puts a lot of stress on the GC  (in 1.11)

![grafik](https://github.com/cesmix-mit/LAMMPS.jl/assets/32484985/1301457a-7f91-486c-8434-24d2358d4b9a)

after running my test loop (with 10^8 iterations) i still need manually trigger GC twice for the memory to be freed completely.

![grafik](https://github.com/cesmix-mit/LAMMPS.jl/assets/32484985/b0348dba-5925-4632-9d03-1fd9c908fe59)

this doesn't happen if I attach the finalizer to the array instead of the memory